test/common/upstream/upstream_impl_test.cc

```diff
@@ -1542,17 +1542,33 @@ TEST_F(ClusterInfoImplTest, ExtensionProtocolOptionsForUnknownFilter) {
       no_such_filter: { option: value }
   )EOF";
 
-  EXPECT_THROW_WITH_MESSAGE(makeCluster(yaml), EnvoyException,
-                            "Didn't find a registered implementation for name: 'no_such_filter'");
+  EXPECT_THROW_WITH_MESSAGE(
+      makeCluster(yaml), EnvoyException,
+      "Didn't find a registered network or http implementation for name: 'no_such_filter'");
 }
 
-class TestFilterConfigFactory : public Server::Configuration::NamedNetworkFilterConfigFactory {
+class TestFilterConfigFactoryBase {
 public:
-  TestFilterConfigFactory(
+  TestFilterConfigFactoryBase(
       std::function<ProtobufTypes::MessagePtr()> empty_proto,
       std::function<Upstream::ProtocolOptionsConfigConstSharedPtr(const Protobuf::Message&)> config)
       : empty_proto_(empty_proto), config_(config) {}
 
+  ProtobufTypes::MessagePtr createEmptyProtocolOptionsProto() { return empty_proto_(); }
+  Upstream::ProtocolOptionsConfigConstSharedPtr
+  createProtocolOptionsConfig(const Protobuf::Message& msg) {
+    return config_(msg);
+  }
+
+  std::function<ProtobufTypes::MessagePtr()> empty_proto_;
+  std::function<Upstream::ProtocolOptionsConfigConstSharedPtr(const Protobuf::Message&)> config_;
+};
+
+class TestNetworkFilterConfigFactory
+    : public Server::Configuration::NamedNetworkFilterConfigFactory {
+public:
+  TestNetworkFilterConfigFactory(TestFilterConfigFactoryBase& parent) : parent_(parent) {}
```

Comment:
It's a little odd not to use inheritance here, but I guess the alternative is assigning the lambdas to variables in the tests and re-using them.

It's a bit dodgy, but I think you could combine the two factories into a single class as long as you overload all the methods that appear in both. 😬

Reply:
The catch is that i'm not sure how that will work with `name()` method that's shared on both, but doesn't come from a common interface. is there a way to express that in c++ ?

Reply:
It's probably a bad idea. I'm fine for you to leave this as is.